hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientResponseHandlerSupplier.java

```diff
@@ -0,0 +1,205 @@
+    }
+
+    private void process(ClientConnection connection, ClientMessage message) {
+        try {
+            handleClientMessage(message);
+        } catch (Exception e) {
+            logger.severe("Failed to process task: " + new ClientPacket(connection, message)
+                    + " on responseThread: " + Thread.currentThread().getName(), e);
+        } finally {
+            connection.decrementPendingPacketCount();
+        }
+    }
+
+    private void handleClientMessage(ClientMessage clientMessage) {
+        long correlationId = clientMessage.getCorrelationId();
+
+        ClientInvocation future = invocationService.deRegisterCallId(correlationId);
+        if (future == null) {
+            logger.warning("No call for callId: " + correlationId + ", response: " + clientMessage);
+            return;
+        }
+
+        if (ErrorCodec.TYPE == clientMessage.getMessageType()) {
+            future.notifyException(client.getClientExceptionFactory().createException(clientMessage));
+        } else {
+            future.notify(clientMessage);
+        }
+    }
+
+    private class ResponseThread extends Thread {
+        private final BlockingQueue<ClientPacket> responseQueue;
+
+        ResponseThread(String name) {
+            super(name);
+            setContextClassLoader(client.getClientConfig().getClassLoader());
+            this.responseQueue = new MPSCQueue<ClientPacket>(this, getIdleStrategy(client.getProperties(), IDLE_STRATEGY));
+        }
+
+        @Override
+        public void run() {
+            try {
+                doRun();
+            } catch (OutOfMemoryError e) {
+                onOutOfMemory(e);
+            } catch (Throwable t) {
+                invocationService.invocationLogger.severe(t);
+            }
+        }
+
+        private void doRun() {
+            while (!invocationService.isShutdown()) {
+                ClientPacket task;
+                try {
+                    task = responseQueue.take();
+                } catch (InterruptedException e) {
+                    continue;
+                }
+                process(task.connection, task.message);
+            }
+        }
+    }
+
+    private class SyncResponseHandler implements ClientResponseHandler {
+        @Override
+        public void handle(ClientMessage message, ClientConnection connection) {
+            process(connection, message);
+        }
+    }
+
+    private class AsyncSingleThreadedResponseHandler implements ClientResponseHandler {
+        @Override
+        public void handle(ClientMessage message, ClientConnection connection) {
+            responseThreads[0].responseQueue.add(new ClientPacket(connection, message));
+        }
+    }
+
+    private class AsyncMultiThreadedResponseHandler implements ClientResponseHandler {
+        @Override
+        public void handle(ClientMessage message, ClientConnection connection) {
+            int threadIndex = INT_HOLDER.get().getAndInc() % responseThreads.length;
```

Comment:
As an optimization and load distribution, what about putting into the queue of a thread with least number of waiting jobs? Would this be too costly to determine which thread has the minimum queue size?

Reply:
This might be a future optimization. But for the time being I would like to keep things simple and would only like to add load balancing if there is a proven performance increase.